service/storage_proxy.cc

```diff
@@ -380,8 +410,17 @@ class storage_proxy::remote {
             tracing::trace(trace_state_ptr, "Message received from /{}", src_addr.addr);
         }
 
-        std::vector<frozen_mutation_and_schema> mutations;
         auto timeout = *t;
+        if (fence && !fms.empty()) {
+            const auto& fm = fms[0];
+            auto stale_topology = co_await try_apply_fence(*fence, timeout);
+            if (stale_topology) {
+                // todo: do it right
+                throw *stale_topology;
+            }
+        }
+
+        std::vector<frozen_mutation_and_schema> mutations;
```

Comment:
Counters have a two-level scheme (see call to mutate_counters_on_leader()). We need to extend fencing to the leader, so it can extend it again to the replica.

Materialized views are similar: coordinator -> replica -> view replica.